### ISSUE TITLE
Possible leaks/out of bounds access in mimalloc 2.1.1

### ISSUE THREAD

**Issue by vkushwaha-nv** [microsoft/mimalloc#731]
3 possible leaks:

1. In _mi_prim_commit(), at 
return (p == addr ? 0 : (int)GetLastError());
Should p be freed when p != addr?

2. Similarly in _mi_prim_reset(), should p be freed when p != addr?

3. FreeLibrary after loading psapi.dll in _mi_prim_process_info() ?

Out of bounds access:
In _mi_segment_huge_page_reset(), if mi_usable_size(block) return 0, wouldn't csize become a very large value that could lead to out of bounds access?

**Comment by daanx** [comment]
Nice! Thank you for the feedback. VirtualAlloc actually always returns NULL on error so I clarified the return value. `FreeLibrary` in process info should not be called as that might make the proc address invalid (if the process didn't use psapi.dll in the first place). The other FreeLibrary calls are used for system libraries that will always be there (but these could be left out as well, I'll ponder it).

The segment huge size underflow is well spotted :-) I fixed it (although I think it will never happen at runtime as this is only used for huge pages, which surely are larger than 8 bytes).
